plutus-metatheory/src/Builtin.lagda.md

```diff
@@ -79,7 +95,127 @@ data Builtin : Set where
   mkPairData                      : Builtin
   mkNilData                       : Builtin
   mkNilPairData                   : Builtin
+```
+
+## Signatures
+
+The following module defines a `signature` function assigning to each builtin an abstract type.
+
+```
+private module SugaredSignature where
+```
+Syntactic sugar for writing the signature of built-ins.
+This is defined in its own module so that these definitions are not exported.
+
+```
+    -- number of different type variables
+    ∙ = 0
+    ∀a = 1
+    ∀b,a = 2
+
+    -- shortened names for type constants and type constructors
+    integer bool bytestring string unit Data : ∀{n} → n ⊢♯
+    integer = con T.integer
+    bool = con T.bool
+    bytestring = con T.bytestring
+    string = con T.string
+    unit = con T.unit
+    Data = con T.Data
 
+    pair : ∀{n} → n ⊢♯ → n ⊢♯ → n ⊢♯
+    pair x y = con (T.pair x y)
+
+    list : ∀{n} → n ⊢♯ → n ⊢♯
+    list x = con (T.list x)
+
+    -- names for type variables
+    a : ∀{n} → suc n ⊢♯
+    a = ` Z
+
+    b : ∀{n} → suc (suc n) ⊢♯
+    b = ` (S Z)
+
+   -- syntax for writing functions of 1,2,3, and 6 arguments
+    _[_]⟶_ : (n : ℕ) → n ⊢♯ → n ⊢♯ → Sig
+    n [ t ]⟶ r = sig n [ t ] r
+
+    _[_,_]⟶_ : (n : ℕ) → n ⊢♯ → n ⊢♯ → n ⊢♯ → Sig
+    n [ t , u ]⟶ r = sig n (t ∷⁺ [ u ]) r 
+
+    _[_,_,_]⟶_ : (n : ℕ) → n ⊢♯ → n ⊢♯ → n ⊢♯ → n ⊢♯ → Sig
+    n [ t , u , v ]⟶ r = sig n (t ∷⁺ u  ∷⁺ [ v ]) r 
+
+    _[_,_,_,_,_,_]⟶_ : (n : ℕ) → n ⊢♯ → n ⊢♯ → n ⊢♯ → n ⊢♯ → n ⊢♯ → n ⊢♯ → n ⊢♯ → Sig
+    n [ t , u , v , w , x , y ]⟶ r = sig n (t ∷⁺ u  ∷⁺ v ∷⁺ w  ∷⁺ x  ∷⁺ [ y ]) r
```

Comment:
I bet you can make this arity-generic while preserving the exact syntax.

Reply:
Perhaps, but I didn't want to spend too much time on this.

Reply:
Now I've done it.